### PR TITLE
Do not remove leading zeros in trace ID hex string

### LINKED ISSUE #1578: [jaeger-query] Leading zeros stripped from trace IDs
## Requirement - what kind of business use case are you trying to solve?

I would like to view traces in Jaeger UI where the trace ID has leading 0s.

## Problem - what in Jaeger blocks you from solving the requirement?

When jaeger-query returns data for traces where there are leading 0s in the trace ID, the leading 0s are stripped from the trace ID, within the data. E.g. for a trace with ID `0123`, all of the spans returned refer to trace ID `123`.

For each trace returned by jaeger-query, this applies to:

* The trace ID on the root of the trace data object
* The trace ID fields in each span
* The trace ID fields in internal references 

## Proposal - what do you suggest to solve the problem or improve the existing situation?

The zeros should not be stripped.

### PR DESCRIPTION
**Pull request by yurishkuro**
Resolves #1578

## Motivation

Among randomly generated and uniformly distributed trace IDs, only 1/16th of them start with 0 followed by a significant digit, 1/256th start with two 0s, and so on in decreasing geometric progression. Therefore, trimming the leading 0s is a very modest optimization on the size of the data being transmitted or stored.

However, trimming 0s leads to ambiguities when the IDs are used as correlations with other monitoring systems, such as logging, that treat the IDs as opaque strings and cannot establish the equivalence between padded and unpadded IDs. It is also incompatible with W3C Trace Context and Zipkin B3 format, both of which include all leading 0s, so an application instrumented with OpenTelemetry SDKs may be logging different trace ID strings than application instrumented with Jaeger SDKs (related issue #1657).

Open question: this change may prevent ES backend from reading the old traces if their IDs had leading zeroes, because previously the zeroes were not added to string representation, and now they are. Roughly one in 16 traces will be affected.

### CODE REVIEW

**Comment by yurishkuro** [comment]
@pavolloffay please review, and any thoughts on the ES question?  One option is to add a backwards compatibility mode by doing a fallback search with unpadded trace ID strings if full-length strings are not found. It would be an option people can configure.

**Comment by yurishkuro** [comment]
cc @jpkrohling (I think Pavol is on PTO)

**Comment by pavolloffay** [comment]
I will have a look at this one, hod for the moment

**Comment by pavolloffay** [comment]
> Open question: this change may prevent ES backend from reading the old traces if their IDs had leading zeroes, because previously the zeroes were not added to string representation, and now they are. Roughly one in 16 traces will be affected.

Indeed, spans previously stored with with id like `1` are not searchable after this change. We definitely need a way to be backward compatible.

**Comment by pavolloffay** [comment]
Something like this should work https://github.com/yurishkuro/jaeger/pull/1/files it needs some tests though.

@yurishkuro could you please summarize why adding leading zeros is necessary?

**Comment by yurishkuro** [comment]
@pavolloffay I added motivation to description.

**Comment by pavolloffay** [comment]
thanks @yurishkuro. It makes sense. I have updated my PR to your branch.

File: plugin/storage/es/spanstore/reader.go [hunk 1]
```diff
@@ -393,15 +393,37 @@ func (s *SpanReader) multiRead(ctx context.Context, traceIDs []model.TraceID, st
...
+	}
+	return elastic.NewBoolQuery().Should(
+		elastic.NewTermQuery(traceIDField, traceIDStr).Boost(2),
+		elastic.NewTermQuery(traceIDField, legacyTraceID))
+}
+
 func convertTraceIDsStringsToModels(traceIDs []string) ([]model.TraceID, error) {
-	traceIDsModels := make([]model.TraceID, len(traceIDs))
-	for i, ID := range traceIDs {
+	traceIDsMap := map[model.TraceID]bool{}
```
Comment (pavolloffay) [inline plugin/storage/es/spanstore/reader.go]: I forgot to add a TODO comment here similar to one in `buildTraceByIDQuery`

Comment (yurishkuro) [reply plugin/storage/es/spanstore/reader.go]: I'll add it. Please approve if you're otherwise ok with the change.

**Comment by pavolloffay** [comment]
We should list this in breaking changes as it might impact integrations

**Pull request merged**